File: go/protobuf/protobuf.go
//go:generate protoc -I=../../schemas --go_out=.. ../../schemas/message.proto

package protobuf

import (
	"log"

	"github.com/kcchu/buffer-benchmarks/constants"
	"github.com/kcchu/buffer-benchmarks/protobuf/generated"
	"google.golang.org/protobuf/proto"
)

func Encode() []byte {
	md := &generated.MessageData{
		Body: &generated.MessageData_CastAddBody{
			CastAddBody: &generated.CastAddBody{
				Text: constants.SampleText,
			},
		},
		Type:      generated.MessageType_CastAdd,
		Timestamp: constants.SampleTimestamp,
		Fid:       constants.SampleFid,
		Network:   generated.FarcasterNetwork_Devnet,
	}
	mdBytes, err := proto.Marshal(md)
	if err != nil {
		log.Fatalln("Failed to encode MessageData:", err)
	}

	m := &generated.Message{
		Data:            mdBytes,
		Hash:            constants.SampleHash,
		HashScheme:      generated.HashScheme_Blake3,
		Signature:       constants.SampleSignature,
		SignatureScheme: generated.SignatureScheme_Ed25519,
		Signer:          constants.SampleSigner,
	}
	mBytes, err := proto.Marshal(m)
	if err != nil {
		log.Fatalln("Failed to encode Message:", err)
	}
	return mBytes
}

func Decode(buf []byte) {
	m := new(generated.Message)
	if err := proto.Unmarshal(buf, m); err != nil {
		log.Fatalln("Failed to decode Message", err)
	}
	md := new(generated.MessageData)
	if err := proto.Unmarshal(m.Data, md); err != nil {
		log.Fatalln("Failed to decode MessageData", err)
	}
	castAddBody := md.Body.(*generated.MessageData_CastAddBody).CastAddBody
	if castAddBody.Text != constants.SampleText {
		log.Fatalln("Unexpected decoded text")
	}
}

File: rust/src/protobuf.rs
mod codegen {
    include!(concat!(env!("OUT_DIR"), "/protobuf/mod.rs"));
}

use protobuf::Message;

use self::codegen::message as generated;
use crate::constants;

pub fn encode() -> Vec<u8> {
    let md = generated::MessageData {
        body: Some(generated::message_data::Body::CastAddBody(
            generated::CastAddBody {
                embeds: Default::default(),
                mentions: Default::default(),
                text: String::from(constants::SAMPLE_TEXT),
                parent: Some(generated::cast_add_body::Parent::CastId(
                    generated::CastId {
                        fid: Vec::from(constants::SAMPLE_FID),
                        ts_hash: Vec::from(constants::SAMPLE_TS_HASH),
                        special_fields: Default::default(),
                    },
                )),
                special_fields: Default::default(),
            },
        )),
        type_: generated::MessageType::CastAdd.into(),
        timestamp: constants::SAMPLE_TIMESTAMP,
        fid: Vec::from(constants::SAMPLE_FID),
        network: generated::FarcasterNetwork::Devnet.into(),
        special_fields: Default::default(),
    };
    let md_buf = md.write_to_bytes().unwrap();

    let m = generated::Message {
        data: md_buf,
        hash: Vec::from(constants::SAMPLE_HASH),
        hash_scheme: generated::HashScheme::Blake3.into(),
        signature: Vec::from(constants::SAMPLE_SIGNATURE),
        signature_scheme: generated::SignatureScheme::Ed25519.into(),
        signer: Vec::from(constants::SAMPLE_SIGNER),
        special_fields: Default::default(),
    };
    m.write_to_bytes().unwrap()
}

pub fn decode(buf: &[u8]) {
    let m = generated::Message::parse_from_bytes(buf).unwrap();
    let md = generated::MessageData::parse_from_bytes(m.data.as_ref()).unwrap();
    let generated::message_data::Body::CastAddBody(body) = md.body.unwrap();
    if body.text != constants::SAMPLE_TEXT {
        panic!("Unexpected decoded text")
    }
}
